src/components/DeviceInteraction/DeviceInteractionStep.js

```diff
@@ -0,0 +1,221 @@
+// @flow
+
+import React, { PureComponent } from 'react'
+
+import Box from 'components/base/Box'
+import { delay } from 'helpers/promise'
+
+import {
+  DeviceInteractionStepContainer,
+  SpinnerContainer,
+  IconContainer,
+  SuccessContainer,
+  ErrorDescContainer,
+  ErrorContainer,
+} from './components'
+
+export type Step = {
+  id: string,
+  title: string | (any => string),
+  desc: any,
+  icon: React$Element<*> | null,
+  run?: any => Promise<any> | { promise: Promise<any>, unsubscribe: void => any },
+  render?: (
+    { onSuccess: any => any, onFail: any => any, onRetry: any => any },
+    any,
+  ) => React$Element<*>,
+  minMs?: number,
+}
+
+type Status = 'idle' | 'running'
+
+type Props = {
+  isFirst: boolean,
+  isLast: boolean,
+  isActive: boolean,
+  isPrecedentActive: boolean,
+  isError: boolean,
+  isSuccess: boolean,
+  isPassed: boolean,
+  step: Step,
+  error: ?Error,
+  onSuccess: (any, Step) => any,
+  onFail: (Error, Step) => any,
+  onRetry: void => any,
+  data: any,
+}
+
+class DeviceInteractionStep extends PureComponent<
+  Props,
+  {
+    status: Status,
+  },
+> {
+  static defaultProps = {
+    data: {},
+  }
+
+  constructor(props: Props) {
+    super(props)
+    const { isFirst } = this.props
+    if (isFirst) {
+      // cf: __IS_MOUNTED__THX_FOR_REMOVING_COMPONENTWILLMOUNT__
+      this.state.status = 'running'
+
+      this.run()
+    }
+  }
+
+  state = {
+    status: 'idle',
+  }
+
+  componentDidMount() {
+    this.__IS_MOUNTED__THX_FOR_REMOVING_COMPONENTWILLMOUNT__ = true
+  }
+
+  componentDidUpdate(prevProps: Props) {
+    const { isActive, error } = this.props
+    const { status } = this.state
+
+    const didActivated = isActive && !prevProps.isActive
+    const didDeactivated = !isActive && prevProps.isActive
+    const stillActivated = isActive && prevProps.isActive
+    const didResetError = !error && !!prevProps.error
+
+    if (didActivated && status !== 'running') {
+      this.run()
+    }
+
+    if (didResetError && stillActivated) {
+      this.run()
+    }
+
+    if (didDeactivated && status === 'running') {
+      this.cancel()
+    }
+  }
+
+  componentWillUnmount() {
+    if (this._unsubscribe) {
+      this._unsubscribe()
+    }
+    this._unmounted = true
+  }
+
+  __IS_MOUNTED__THX_FOR_REMOVING_COMPONENTWILLMOUNT__ = false
+  _unsubscribe = null
+  _unmounted = false
+
+  handleSuccess = (res: any) => {
+    const { onSuccess, step, isError } = this.props
+    if (isError) return
+    this.safeSetState({ status: 'idle' })
+    onSuccess(res, step)
+  }
+
+  handleFail = (e: Error) => {
+    const { onFail, step } = this.props
+    this.safeSetState({ status: 'idle' })
+    onFail(e, step)
+  }
+
+  safeSetState = (...args: any) => {
+    if (this._unmounted) return
+    this.setState(...args)
+  }
+
+  run = async () => {
+    const { step, data } = this.props
+
+    if (this.__IS_MOUNTED__THX_FOR_REMOVING_COMPONENTWILLMOUNT__) {
+      this.safeSetState({ status: 'running' })
```

Comment:
i would prefer not to use this `safeSetState` pattern. this is a hack to hide the bigger problem^^ we don't want to silent react errors.

& we really want to stop the run logic if it's unmounted

second reason is safeSetState does not help flow typecheck.

```
if (this._unmounted) return
this.setState(...args)
```